WIP: markdown: View code in playground option for codeblocks

sumanthvrao opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This is a prototype PR to solve #11618 and right now hardcodes the playground link for ~5 languages (Rust, python2, python3, Lean, Julia). Post this we can start working on supporting Realm Admins configuring their own choices for playground links.

**Testing Plan:** <!-- How have you tested? -->
Manually testing done to verify that clicking on the link opens the playground (and loads the code).

**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![view-code-2](https://user-images.githubusercontent.com/30312566/94343448-efb7d100-0035-11eb-96dd-f1b7844f35a6.gif)


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->
Fixes: #11618 

## TODO:

- [x] Tweak build_pygments_data tool to generate `pygments_lang.json` in a format most useful for us.
- [x] We should add a mention of the new markdown feature in `templates/zerver/api/changelog.md` and bump the feature level.
- [x] Experiment with a dropdown UI for multiple playground options.
- [ ] Start working towards adding support for realm admins to configure their own playground choices (open up an issue for this)
- [ ] Add Help Center documentation for this once we've properly documented the configuration interface.

showell commented:
minor typo in commit message--s/mimicks/mimics (there's no "k")

showell commented on zerver/lib/markdown/fenced_code.py:
```diff
@@ -392,6 +393,13 @@ def format_code(self, lang: str, text: str) -> str:
+        # CodeHilite generates to add this language as a data-attribute.
+        if lang:
+            parsed_code = BeautifulSoup(code, features='lxml')
+            parsed_code.find('div')['data-codehilite-language'] = lang
```
We should probably add some comments here that make it a bit more clear where that div is coming from.  It's not coming from CodeHilite, is it?  Here is the relevant code from CodeHilite: https://github.com/Python-Markdown/markdown/blob/master/markdown/extensions/codehilite.py#L158

timabbott replied:
I think `self.use_pygments` is true and it doesn't use that code path.  See the `css_class` option docs for some detail on how this is configured.

timabbott commented on zerver/lib/markdown/fenced_code.py:
```diff
@@ -392,6 +393,13 @@ def format_code(self, lang: str, text: str) -> str:
+        # we need to know the language used in the codeblock. We tweak the HTML
+        # CodeHilite generates to add this language as a data-attribute.
+        if lang:
+            parsed_code = BeautifulSoup(code, features='lxml')
```
Do we need BeautifulSoup here?  I'd have expected that we'd be able to use LXML directly without a `find` step, since the outer element of `code` is known to be the `div`.

timabbott commented on zerver/tests/test_markdown.py:
```diff
@@ -1371,11 +1371,12 @@ def test_default_code_block_language(self) -> None:
 
         # Render without default language
-        msg_with_js = markdown_convert_wrapper(text.format('js'))
+        msg_with_js = markdown_convert_wrapper(text.format('javascript'))
```
Why did you make this change?  It looks like we were using an alias `js` for `javascript` before; I suspect that we should not have removed that test and instead changed the implementation to canonicalize aliases like this when generating `data-codehilite-class`?

timabbott commented on static/js/settings_config.js:
```diff
@@ -320,3 +320,17 @@ exports.all_notifications = () => ({
+
+exports.get_playground_for_languages = function (lang) {
+    // TODO: This is being hardcoded for now just for the prototype, post which we should
+    // add support for realm admins to configure their own choices.
```
Just as as note on what we'll likely want as the eventual configuration API, I think we want each language to map to a list of objects, each object having a `name` and `url_pattern` fields, with `url_pattern` having a `{code}` parameter in it to substitute.

timabbott commented on static/js/rendered_markdown.js:
```diff
@@ -190,10 +192,28 @@ exports.update_elements = (content) => {
-    // Display the copy-to-clipboard button inside the div.codehilite element.
+    // Display the view-code-in-playground and the copy-to-clipboard button inside the div.codehilite element.
     content.find("div.codehilite").each(function () {
+        const $codehilite = $(this);
```
It might be readable to accept an argument to the function `$elt` rather than using `$(this)`.

sumanthvrao replied:
I didn't understand what you meant by this.

timabbott replied:
Rather than using `this`, access the `div.codehilite` element that jQuery passed into this function (which we're currently ignoring).

timabbott commented:
This is great; posted a batch of comments.  One thing we should think about is the UI pattern we want; I kinda like the 1-button model you have, but there may be situations where we let users configure multiple options (E.g. "run in Rust 1.0" vs. "Run in Rust 2.0"), where a dropdown UI might be appropriate.  

I'll post on the issue thread for feedback on how real a concern this is.

sumanthvrao commented:
@showell @timabbott Thanks for the review. Here's a changelog of the update I pushed

* Changed BeautifulSoup parser to lxml.
* Also changed the implementation to handle canonicalized alias (with the approach mentioned by Anders on czo).
* Also changed the code-extraction logic to be on-click of the view-in-playground button (rather than passing the extracted code during button creation itself)

Couple of things I am unsure of (hence the WIP) - Majorly the last two commits
* I think we need the canonical-alias mappings passed to the frontend for things like displaying the admin configuration setting (future) or even for the local echo logic - by tweaking the tool generating `pygments_data.json` (see last but one commit); But this seems like a lot of duplication, so we should change the old structure of lang-code mapping to add this info there itself. What do you suggest?
* Also we would need to parse this generated json to mimic the backend logic for local echos. (See last commit). Is that okay?

timabbott commented:
> But this seems like a lot of duplication, so we should change the old structure of lang-code mapping to add this info there itself. What do you suggest?

I would recommend changing `pygments_data.json` to have data in the format we want; that's a file we control entirely and is intended for this exact purpose.  

> Also we would need to parse this generated json to mimic the backend logic for local echos. (See last commit). Is that okay?

Yeah, that should be OK.

timabbott commented on zerver/lib/markdown/fenced_code.py:
```diff
@@ -392,6 +395,24 @@ def format_code(self, lang: str, text: str) -> str:
+            div_tag.attrib['data-codehilite-language'] = lexer_subclass_name
+            # Lxml implicitly converts tags like <span></span> into <span/>
+            # specifying method="c14n" when converting to string, prevents that.
+            code = etree.tostring(div_tag, method="c14n").decode()
```
It would be kinda nice if the pygments offered a way to specify attributes (not just a list of CSS classes) to attach to the `pre` element being generated by `pygments`; this might require upstream contributions to both `pygments` and then `python-markdown`.  @andersk FYI in case you have ideas; I don't see a better way to do what we want with their current API.

timabbott commented on zerver/tests/fixtures/markdown_test_cases.json:
```diff
@@ -3,8 +3,8 @@
-      "expected_output": "<p>Hamlet said:</p>\n<div class=\"codehilite\"><pre><span></span><code><span class=\"k\">def</span> <span class=\"nf\">speak</span><span class=\"p\">(</span><span class=\"bp\">self</span><span class=\"p\">):</span>\n    <span class=\"n\">x</span> <span class=\"o\">=</span> <span class=\"mi\">1</span>\n</code></pre></div>",
-      "marked_expected_output": "<p>Hamlet said:</p>\n<div class=\"codehilite\"><pre><span></span><code>def speak(self):\n    x = 1\n</code></pre></div>",
+      "expected_output": "<p>Hamlet said:</p>\n<div class=\"codehilite\" data-codehilite-language=\"Python\"><pre><span></span><code><span class=\"k\">def</span> <span class=\"nf\">speak</span><span class=\"p\">(</span><span class=\"bp\">self</span><span class=\"p\">):</span>\n    <span class=\"n\">x</span> <span class=\"o\">=</span> <span class=\"mi\">1</span>\n</code></pre></div>",
+      "marked_expected_output": "<p>Hamlet said:</p>\n<div class=\"codehilite\" data-codehilite-language=\"Python\"><pre><span></span><code>def speak(self):\n    x = 1\n</code></pre></div>",
```
This commit shouldn't include the `data-codehilite` bit in `marked_expected_output`; if you fix that we can merge the first commit.

timabbott commented on static/templates/view_code_in_playground.hbs:
```diff
@@ -0,0 +1 @@
+<a target="_blank" class="code_external_link"  title="{{title}}" aria-label="{{title}}"><i class="fa fa-external-link" aria-hidden="true"></i></a>
```
I don't think this should be aria-hidden - -that makes it inaccessible.  aria-hidden should be used only for elements that aren't clickable (E.g. the decoration icons we have next to text in some places)

timabbott replied:
Also this needs `rel=noopener` type tags for security; look at any of our other links for an example.

sumanthvrao replied:
Interesting... Thanks!

timabbott commented:
I merged the first commit.  Thanks @sumanthvrao!

timabbott commented:
I just pushed back to this PR with a new commit added; I was originally planning to just improve comments, but working on that made me realize there's a corner case we haven't discussed -- what to do with a language that doesn't appear in `Pygments`.  Do we want to allow folks to specify any made-up language name they want?  It would provide a "bonus feature" of letting folks specify a "playground" for arbitrary internal languages (or variants like `rust2022`) that might not have syntax highlighting.

But if we do that, we might want to rename `data-codehilite-language` to the simpler `data-code-language`, which might be a better name anyway.

timabbott commented:
@sumanthvrao want to add a commit at the bottom of the branch to do the `data-code-language` rename?  I think we likely want to do that in any case.

timabbott commented:
A few other notes:
* We should add a mention of the new markdown feature in `templates/zerver/api/changelog.md` and bump the feature level.
* I suppose we can't do Help Center documentation for this until we've properly documented the configuration interface.

sumanthvrao commented:
Thanks! Pushed a couple of updates (minor ones) - tag rename and your cleanup commit;
>  Do we want to allow folks to specify any made-up language name they want? It would provide a "bonus feature" of letting folks specify a "playground" for arbitrary internal languages (or variants like `rust2022`) that might not have syntax highlighting.

Seems pretty useful to be able to configure something like this. (Hence my original commit just echoed back the text entered)

timabbott commented:
Great.  I merged the first two commits as ae58ed5a74d9fb902bc5f746086cf0400009abcf after adjusting for our discussion.

timabbott commented on tools/setup/build_pygments_data:
```diff
@@ -14,8 +14,11 @@ with open(DATA_PATH) as f:
+        pygments_language_code = 0
+        if name in langs:
+            pygments_language_code = langs[name]
+        langs[name] = [pygments_language_code, pygments_language_name]
```
I think it'd be nicer to just do an object with `code: ` and `name: ` attributes; that'll result in cleaner code accessing this (one doesn't have to remember which thing in [0] vs. [1] if it's accessed by name).

timabbott replied:
Also, what you've called `code` should be called `priority` -- it's used to decide sort order.  So maybe

```
langs[name] = {
    'priority': langs.get(name, 0),
    'pretty_name': get_lexer_by_name(name).name,
}
```

timabbott replied:
This commit should also add 1 to PROVISION_VERSION, since one needs to provision following this change.

sumanthvrao replied:
Sounds good :+1:

timabbott commented on static/shared/js/fenced_code.js:
```diff
@@ -30,14 +33,21 @@ let stash_func = function (text) {
+    // to know what Pygments language was used to highlight this code block.
+    if (lang !== undefined && lang !== "") {
+        const pygments_language_info = pygments_data.langs[lang];
+        let code_language = lang;
```
Do we need to HTML-escape `code_language`?  Might be best to do it or add a comment on what it's not necessary.

timabbott commented on static/js/settings_config.js:
```diff
@@ -320,3 +320,36 @@ exports.all_notifications = () => ({
+    Lean: {
+        name: "Lean playground",
+        url_pattern: "https://leanprover-community.github.io/lean-web-editor/#code=",
+    },
```
I think format-wise, we should make each of these a list?

sumanthvrao replied:
Yup! This was my next step :)

sumanthvrao commented:
Pushed an update addressing the feedback for first two commits - I added a new test in the second commit (might be useful since we don't have any test canonicalizing the language alias); I am yet to fix the hardcoded format and adding the UI, etc (third commit and onwards) - Hope to work on that soon :)

timabbott commented:
I merged the first 4 commits as the series ending with 2b9f7916c57df9daf02815d265649f37a7ad4b2e, after adding an extra bit of comment alongside the escaping for clarity.

What are you thinking for the UI?  One model that could be nice is to display two icons: Copy and View In Playground.  The latter would be a 1-click UX if there's only a single playground configured for that language, and open a popover showing the options otherwise.

sumanthvrao commented:
Thanks! Yup, that design for the UI sounds good. I'll try posting some early design looks (soon) on the czo thread for more feedback.

andersk reviewed (changes requested):
These lint errors that you disabled in c563cdba61544101ad52397770da0eaf5a75e8da were there for a reason.

```diff
--- a/static/shared/js/fenced_code.js
+++ b/static/shared/js/fenced_code.js
@@ -1,6 +1,9 @@
 import katex from "katex";
 import _ from "lodash";
 
+// eslint-disable-next-line
+const pygments_data = require("../../generated/pygments_data.json");
+
 // Parsing routine that can be dropped in to message parsing
 // and formats code blocks
 //
```

```
/srv/zulip/static/shared/js/fenced_code.js
  4:23  error  'require' is not defined                                                     no-undef
  4:31  error  Unexpected path "../../generated/pygments_data" imported in restricted zone  import/no-restricted-paths
```

The `@zulip/shared` package cannot use files outside of itself, otherwise it will not work in zulip-mobile.

Also, `require` does not generally work in ES6 modules.

Also, never disable ESLint completely, only specific rules: `// eslint-disable-next-line no-undef, import/no-restricted-paths`.

sumanthvrao commented:
Thanks for pointing these out. I'm not entirely sure what the right solution here is
* `require` can be replaced by an import line and that should work.
* However, this still leaves us with accessing the `/generated` folder for pygments data, outside the `/shared` folder. 

(I suppose there is the option of marking this as `backend_only_syntax` as a last resort)

andersk commented:
If `shared` uses `pygments_data`, it needs to receive it via function argument, similarly to how `emoji.initialize` receives `emoji_codes`.

sumanthvrao commented:
The first commit here has the fix to the issue Anders mentioned above. 

FYI @timabbott

sumanthvrao commented:
Pushed a bunch of commits here which adds the UI model for the "view in playground" option. This is ready for a review.

timabbott commented:
Merged as 87c809c0e31246b0cb4f737d80c1629a7aeba495, after several small i18n-related tweaks.  Huge thanks for building this @sumanthvrao!  I opened https://github.com/zulip/zulip/issues/16432 for the follow-up work required to make this a complete feature.
